crates/mun_syntax/src/tests/parser.rs

```diff
@@ -365,3 +365,14 @@ fn type_alias_def() {
     "#,
     )
 }
+#[test]
+fn function_return_path() {
+    snapshot_test(
+        r#"
+        fn main() -> self::Foo {}
```

Comment:
Do we want to support `self` for paths? Based on previous discussion we had set these:
![image](https://user-images.githubusercontent.com/6917585/96120625-1b203400-0eef-11eb-872a-9a31d9abc178.png)

Reply:
`pub(self)` is not valid. 

`self` can only be used in module paths. I use `self` sometimes to make it explicit that the given type or function comes from this module. In the test examples, it is also used to explicitly distinguish the `Foo` coming from this module. I also use it to reexport something from a child module: `pub use self::foo::Bar`

Do you think that the redundancy outweighs the use of the explicitly distinguishing of the current module in paths?

Reply:
I'm not super opinionated about it. I always found it interesting that there were two ways to do the same thing in Rust (although one might be more explicit than the other), as it is likely to cause confusion. "If there are two ways, am I doing it the right way?" 🤔 

For example:
```rust
struct Bar;

mod foo {
  pub struct Bar;
}

use foo::Bar;
```

Doing this results in a compile error:
```bash
   Compiling playground v0.0.1 (/playground)
error[E0255]: the name `Bar` is defined multiple times
 --> src/main.rs:7:5
  |
1 | struct Bar;
  | ----------- previous definition of the type `Bar` here
...
7 | use foo::Bar;
  |     ^^^^^^^^ `Bar` reimported here
  |
  = note: `Bar` must be defined only once in the type namespace of this module
help: you can use `as` to change the binding name of the import
  |
7 | use foo::Bar as OtherBar;
  |     ^^^^^^^^^^^^^^^^^^^^
```

so there is no actual (mandatory) use case for using the `self::Bar`. (I'm assuming here that the argument extends to a use-statement of sub-modules)

Reply:
I get your point. I mainly use it to indicate that something is comming from this or one of the child modules. Mostly in use statements. In `use` statements you always have to start with the package name OR a local module. I use `self` here to be clear that an import name is relative to the current module instead of being a crate. Also if there is a child module with the same name as a package it makes sense to use it (vs `super::<self_name>::foo`)

I do still see some usecases that makes this useful so if you dont mind too much I'll leave it in.

Reply:
I think avoiding conflicts with crate/package names is a very good point.
In that case, is there value to always requiring `self` to avoid this confusion; or just leave it up to developers to decide for themselves?

Reply:
I do see a usecase for that in `use` statements but I wouldnt want to always have to type self there.